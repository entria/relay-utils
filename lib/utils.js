'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.hasPreviousPage = hasPreviousPage;
exports.hasNextPage = hasNextPage;
exports.createDataArray = createDataArray;
exports.identifyKey = identifyKey;

var _get = require('lodash/get');

var _get2 = _interopRequireDefault(_get);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getConnectionData(data, key) {
  if (!data) {
    return {
      edges: [],
      pageInfo: {
        hasPreviousPage: false,
        hasNextPage: false
      }
    };
  }

  return (0, _get2.default)(data, key);
}

function hasPreviousPage(data, key) {
  var identifiedKey = key || identifyKey(data);

  var _getConnectionData = getConnectionData(data, identifiedKey),
      pageInfo = _getConnectionData.pageInfo;

  return (typeof pageInfo === 'undefined' ? 'undefined' : _typeof(pageInfo)) !== (typeof undefined === 'undefined' ? 'undefined' : _typeof(undefined)) && pageInfo.hasPreviousPage;
}

function hasNextPage(data, key) {
  var identifiedKey = key || identifyKey(data);

  var _getConnectionData2 = getConnectionData(data, identifiedKey),
      pageInfo = _getConnectionData2.pageInfo;

  return (typeof pageInfo === 'undefined' ? 'undefined' : _typeof(pageInfo)) !== (typeof undefined === 'undefined' ? 'undefined' : _typeof(undefined)) && pageInfo.hasNextPage;
}

function createDataArray(data, key) {
  var identifiedKey = key || identifyKey(data);

  var _getConnectionData3 = getConnectionData(data, identifiedKey),
      edges = _getConnectionData3.edges;

  return edges.map(function (info) {
    return info.node;
  });
}

function identifyKey(data) {
  if (!data) {
    return null;
  }

  var walkKeys = function walkKeys(props) {
    var key = Object.keys(props)[0];
    var arrayKeys = [key];
    var newProps = props[key];

    if (!newProps.edges) {
      return arrayKeys.concat(walkKeys(newProps));
    }

    return arrayKeys;
  };

  return walkKeys(data).join('.');
}