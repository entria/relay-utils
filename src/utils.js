// @flow
function isObject(data: any): boolean {
  return typeof data === 'object';
}

export function identifyKey(data: Object): ?string {
  if (!data || data === null || Object.keys(data).length === 0) {
    return null;
  }

  const walkProps = (props, parentKeys = []) => {
    const withEdges = Object.keys(props)
      .filter(key => isObject(props[key]))
      .map(key => {
        const newProps = props[key];
        const parentKeysAndMe = parentKeys.concat(key);

        if (typeof newProps.edges !== 'undefined') {
          return parentKeysAndMe.join('.');
        }

        return walkProps(newProps, parentKeysAndMe);
      })
      .filter(key => key !== null);

    return withEdges.length > 0 ? withEdges[0] : null;
  };

  return walkProps(data);
}
