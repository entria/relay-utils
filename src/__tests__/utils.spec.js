import { identifyKey } from '../utils';

it('should return null if data is an empty object', async () => {
  const data = {};

  const identifiedKey = identifyKey(data);

  expect(identifiedKey).toMatchSnapshot();
});

it('should return null if do not has edges', async () => {
  const data = {
    propOne: {
      subPropOne: 'john',
      subPropTwo: 'doe',
    },
    propTwo: {
      subPropThree: 'jane',
      subPropFour: 'doe',
    },
  };

  const identifiedKey = identifyKey(data);

  expect(identifiedKey).toMatchSnapshot();
});

it('should return a property with edges (propTwo)', async () => {
  const data = {
    propOne: {
      subPropOne: 'john',
      subPropTwo: 'doe',
    },
    propTwo: {
      edges: [],
    },
  };

  const identifiedKey = identifyKey(data);

  expect(identifiedKey).toMatchSnapshot();
});

it('should return a subproperty with edges (propTwo.subPropFour)', async () => {
  const data = {
    propOne: {
      subPropOne: 'john',
      subPropTwo: 'doe',
    },
    propTwo: {
      subPropThree: 'jane',
      subPropFour: {
        edges: [],
      },
    },
  };

  const identifiedKey = identifyKey(data);

  expect(identifiedKey).toMatchSnapshot();
});

it('should return the first property with edges (propOne)', async () => {
  const data = {
    propOne: {
      edges: [],
    },
    propTwo: {
      edges: [],
    },
  };

  const identifiedKey = identifyKey(data);

  expect(identifiedKey).toMatchSnapshot();
});
