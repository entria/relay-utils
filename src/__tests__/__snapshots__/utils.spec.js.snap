// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return a property with edges (propTwo) 1`] = `"propTwo"`;

exports[`should return a subproperty with edges (propTwo.subPropFour) 1`] = `"propTwo.subPropFour"`;

exports[`should return null if data is an empty object 1`] = `null`;

exports[`should return null if do not has edges 1`] = `null`;

exports[`should return the first property with edges (propOne) 1`] = `"propOne"`;
